eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+     * {@link MultiSourceMessageStream#hasNextAvailable(int, TimeUnit)} is also configurable.
+     */
+    public static class Builder {
+
+        private Comparator<TrackedSourcedEventMessage<?>> comparator = Comparator.comparing(EventMessage::getTimestamp);
+        private Map<String, StreamableMessageSource<TrackedEventMessage<?>>> messageSourceMap = new LinkedHashMap<>();
+        private String longPollingSource = "";
+
+        /**
+         * Adds a message source to the list of sources.
+         *
+         * @param messageSourceName a unique name identifying the stream.
+         * @param messageSource     the message source to be added.
+         * @return the current Builder instance, for fluent interfacing.
+         */
+        public Builder addMessageSource(String messageSourceName,
+                                        StreamableMessageSource<TrackedEventMessage<?>> messageSource) {
+            Assert.isFalse(messageSourceMap.containsKey(messageSourceName),
+                           () -> "the messageSource name must be unique");
+            Assert.isFalse(messageSourceMap.containsValue(messageSource), () -> "this message source is duplicated");
+
+            messageSourceMap.put(messageSourceName, messageSource);
+            return this;
+        }
+
+        /**
+         * Overrides the default comparator (which returns the oldest event available).
+         *
+         * @param comparator the comparator to use when deciding on which message to return.
+         * @return the current Builder instance, for fluent interfacing.
+         */
+        public Builder withComparator(Comparator<TrackedSourcedEventMessage<?>> comparator) {
+            this.comparator = comparator;
+            return this;
+        }
+
+
+        /**
+         * Select the message source which is most suitable for long polling. To prevent excessive polling on all
+         * sources
+         * it is preferable to do the majority of polling on a single source. All other streams will be checked first
+         * using {@link BlockingStream#hasNextAvailable()} before {@link BlockingStream#hasNextAvailable(int, TimeUnit)}
+         * is
+         * called on the source chosen for long polling. This is then repeated multiple times to increase the chance of
+         * successfully finding a message before the timeout. If no particular source is configured, long polling with
+         * be
+         * done on the last configured source whilst other streams will be polled using {@link
+         * BlockingStream#hasNextAvailable()}
+         *
+         * @param longPollingSource the {@code messageSourceName} on which to do the long polling.
+         * @return the current Builder instance, for fluent interfacing.
+         */
+        public Builder configureLongPollingSource(String longPollingSource) {
+            Assert.isTrue(messageSourceMap.containsKey(longPollingSource),
```

Comment:
I think it makes sense to utilize the `BuilderUtils` here instead of `Assert`, as those ensure we throw a `AxonConfigurationException`.